### PR TITLE
Peregrine implementation/extract view state

### LINKED ISSUE #1600: Extract view state from reducers
View state should not be global, and async actions should not control it. Move view state from reducers and logic from actions into hooks within components. For example, in the user reducer `isSignedIn` is a valid user state but `signInError` is a view concern.

# TODO

* Depending on level of effort this issue may be broken down further per component. Make that call and update.

## State Work

The following is a list of reducers and associated state that I've identified as `view` state. I may have missed some state that is added in the reducer functions but this should be pretty comprehensive.

* app
  * none, afaict. 
* cart
  * `editItem`
  * `isLoading`
  * `isEditingItem`
  * `isUpdatingItem`
  * `isAddingItem`
* catalog
  * none, afaict. 
* checkout
  * `isAddressInvalid`
  * `invalidAddressMessage`
  * `step`
  * `submitting`
* checkoutReceipt
  * none, afaict. 
* directory
  * none, afaict. 
* purchaseDetails
  * `isFetching`
* purchaseHistory
  * `isFetching`
* user
  * `isGettingDetails`
  * `isSigningIn`
  * `forgotPassword.isInProgress`

### PR DESCRIPTION
**Pull request by sirugh**
## Description

This PR moves state relevant to views out of the store and into the components that care about it. See the issue for more information.

## Related Issue

Closes #1600.

## Verification Steps

This touches a _lot_ of code. A regression test should cover our bases on top of unit tests, but I encourage you to try various actions/ buttons within the app that utilize redux to ensure I didn't break anything.

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.

### CODE REVIEW

**Comment by sirugh** [comment]
Based on discussion with @jimbo we decided to keep loading/submitting/fetching states in the global state for now. These are tied directly to an asyncAction's internal side effect and as such are not necessarily related to view state. Though it is possible to do the following, it only makes sense in places where the action is directly tied to the outcome. In other places, such as `getCartDetails` that are widely used by various components, there is no appropriate way to set `isLoading` on _just_ the cart.

File: packages/venia-ui/lib/components/ForgotPassword/forgotPassword.js [hunk 1]
```diff
@@ -9,30 +9,31 @@ import defaultClasses from './forgotPassword.css';
...
+    const [inProgress, setInProgress] = useState(false);
+    const [forgotPasswordEmail, setForgotPasswordEmail] = useState(null);
+
     const classes = mergeClasses(defaultClasses, props.classes);
 
     const handleFormSubmit = useCallback(
-        ({ email }) => {
-            resetPassword({ email });
+        async ({ email }) => {
+            setInProgress(true);
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/components/ForgotPassword/forgotPassword.js]: Ended up creating a state variable like `isResettingPassword` so this is fine.

File: packages/venia-ui/lib/reducers/checkout.js [hunk 2]
```diff
@@ -77,103 +55,39 @@ const reducerMap = {
             }))
         };
     },
-    [actions.shippingAddress.submit]: state => {
-        return {
-            ...state,
-            submitting: true
-        };
-    },
+    [actions.shippingAddress.submit]: state => state,
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/reducers/checkout.js]: Ended up adding in loading states back in so these are all valid again

File: packages/venia-ui/lib/actions/checkout/asyncActions.js [hunk 3]
```diff
@@ -13,7 +13,23 @@ const storage = new BrowserPersistence();
 
 export const beginCheckout = () =>
     async function thunk(dispatch) {
-        dispatch(actions.begin());
+        const storedBillingAddress = storage.getItem('billing_address');
+        const storedPaymentMethod = storage.getItem('paymentMethod');
+        const storedShippingAddress = storage.getItem('shipping_address');
+        const storedShippingMethod = storage.getItem('shippingMethod');
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/actions/checkout/asyncActions.js]: Not that I could see. Only one asyncAction was invoking `actions.begin`.

File: packages/venia-ui/lib/actions/checkout/asyncActions.js [hunk 4]
```diff
@@ -91,33 +107,35 @@ export const submitPaymentMethodAndBillingAddress = payload =>
 
 export const submitBillingAddress = payload =>
     async function thunk(dispatch, getState) {
-        dispatch(actions.billingAddress.submit(payload));
+        dispatch(actions.billingAddress.submit());
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/actions/checkout/asyncActions.js]: Yes. I couldn't tell why we were submitting them other than for logging purposes. If you want me to add them back I certainly can.

File: packages/venia-ui/lib/actions/checkout/asyncActions.js [hunk 5]
```diff
@@ -91,33 +107,35 @@ export const submitPaymentMethodAndBillingAddress = payload =>
...
 
-        const { cart, directory } = getState();
+        const {
+            cart,
+            directory: { countries }
+        } = getState();
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/actions/checkout/asyncActions.js]: Sure. I can separate this.

File: packages/venia-ui/lib/actions/checkout/asyncActions.js [hunk 6]
```diff
@@ -91,33 +107,35 @@ export const submitPaymentMethodAndBillingAddress = payload =>
...
-            } catch (error) {
-                dispatch(actions.billingAddress.reject(error));
-                throw error;
             }
-        }
 
-        await saveBillingAddress(desiredBillingAddress);
-        dispatch(actions.billingAddress.accept(desiredBillingAddress));
+            await saveBillingAddress(desiredBillingAddress);
+            dispatch(actions.billingAddress.accept(desiredBillingAddress));
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/actions/checkout/asyncActions.js]: The only difference would be that we may call `dispatch(actions.billingAddress.reject(error));` with an error from `saveBillingAddress`. If that happens we update the `billingAddressError` payload with the error message from that new potential point of failure where before we would just throw out of the asyncAction.

So the question was whether or not we want an error from `saveBillingAddress` to be propagated to the store, and I think we do.

File: packages/venia-ui/lib/actions/user/asyncActions.js [hunk 7]
```diff
@@ -79,61 +79,44 @@ export const getUserDetails = () =>
...
-        dispatch(actions.resetPassword.request(email));
+        dispatch(actions.resetPassword.request());
 
         // TODO: actually make the call to the API.
         // For now, just return a resolved promise.
-        const response = await Promise.resolve(email);
+        await Promise.resolve(email);
 
-        dispatch(actions.resetPassword.receive(response));
+        dispatch(actions.resetPassword.receive());
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/actions/user/asyncActions.js]: Yes. I removed payloads everywhere they were unused. In this case I refactored how the reset password feature worked so that I didn't have to maintain the submitted password in global state.

File: packages/venia-ui/lib/components/Checkout/Receipt/receipt.js [hunk 8]
```diff
@@ -5,15 +5,16 @@ import Button from '../../Button';
...
 const Receipt = props => {
-    const { createAccount, history, reset, user } = props;
+    const { createAccount, history, reset, onClose, user } = props;
 
     const classes = mergeClasses(defaultClasses, props.classes);
 
     useEffect(() => reset, [reset]);
 
     const handleCreateAccount = useCallback(() => {
+        onClose();
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/components/Checkout/Receipt/receipt.js]: Yea, you're right. I'm not sure why I did this. I'll double check and respond.

File: packages/venia-ui/lib/components/Checkout/flow.js [hunk 9]
```diff
@@ -63,26 +56,42 @@ const Flow = props => {
...
+    }, [beginCheckout, setStep]);
+
+    const handleCancelCheckout = useCallback(async () => {
+        await cancelCheckout();
+        setStep('cart');
+    }, [cancelCheckout, setStep]);
+
+    const handleSubmitOrder = useCallback(async () => {
+        await submitOrder();
+        setStep('receipt');
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/components/Checkout/flow.js]: So I was letting the error throw out of the callback in order to not proceed to the next step.

File: packages/venia-ui/lib/components/Checkout/overview.js [hunk 10]
```diff
@@ -41,6 +41,10 @@ const Overview = props => {
         setEditing('shippingMethod');
     }, [setEditing]);
 
+    const currencyCode =
+        (cart && cart.totals && cart.totals.quote_currency_code) || 'USD';
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/components/Checkout/overview.js]: When we complete a checkout we reset the state of the cart/totals to the default. At that point, _just_ before we transition to the receipt, we render the updated-but-empty overview so it was throwing an error about not passing valid props to `Price`. This fixes that. I could have also tried to update the view state to not render the empty overview but this was much easier.

File: packages/venia-ui/lib/components/Checkout/addressForm.css [hunk 11]
```diff
@@ -51,7 +51,8 @@
     grid-column-end: span 2;
 }
 
-.validation {
+.validationMessage {
     grid-column-end: span 2;
     line-height: normal;
+    color: red;
```
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/components/Checkout/addressForm.css]: Other errors in other forms are red.

File: packages/venia-ui/lib/components/Checkout/Receipt/receipt.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/venia-ui/lib/components/Checkout/Receipt/receipt.js]: Ok so what happens here is that the user clicks `Create Account` and we navigate them, without reloading, to `/create-account`. The order here but it is necessary to call `onClose` so that the next time we render the cart it is on the correct step.

TL;DR - I'll move this after the `createAccount` call.
